This repository contains code to evaluate commonly used computational tools to estimate single-cell copy numbers (CNs) from single-cell whole-genome sequencing data.

### How to setup

```
bash -evx install\_step1\_by\_conda.sh
bash -evx install\_step2\_by\_download\_and\_setup.sh
```

Download the FASTQ files described in scDNAaccessions.tsv (or scDNAaccessions.S04.tsv to perform a test run that is much faster than the full run) into the directory ../data/1from0.datdir/

### How to benchmark the tools

```
python main.py > Snakefile # or python main.py --SraRunTable scDNAaccessions.S04.tsv > Snakefile # to perform the test run
snakemake --cores ${NUM\_CPUS}
```

### Detail about the data

for a set of cells derived from the same donor (e.g., human subject):
    data1 are SRA raw FastQ files
    data2 are reference-aligned BAM files with two germline haplotypes (PB1, PB2 and FPN are separated from each other for oocytes)
    data3 are reference-aligned BAM files with simulated copy numbers
    data4 are single-cell copy number (CN) callling-result TSV and CN benchmark-result TSV files

The file common.py describes the files generated by main.py in more detail

### Other things

Please be aware that the commercial (i.e., for profit and non-academic) use of cosmic-v97/cell\_lines\_copy\_number.csv may require licensing from cancer.sanger.ac.uk/cosmic

### LICENSE

MIT license

